### ISSUE TITLE
Magento2 Attribute option does not validate for existing records before insert

### ISSUE THREAD

**Issue by zeeshan2523806** [magento/magento2#16852]
### Preconditions

When creating product attribute and their options dynamically magento does not validate the data before inserting it into the database,

1. Magento2.2.3
2. 

### Steps to reproduce

1. Create a product eav attribute dynamically using class (Magento\Catalog\Model\ResourceModel\Eav\Attribute) and set the entity type Id to the type id of catalog_product (See sample code below)  (eg: attribute code = colors)
2. generate a list of possible options and then add those options into that product EAV attribute by using (Magento\Eav\Setup\EavSetup -> AddOption()) (optionarray = 'Black, Blue, White')
3. Generate some more options (pick some of those which are already existed) and add then add them into the product attribute by following the same step describe in step 2. ('White, Green, Purple')

### Expected result

1. Magento should validate the option Array and check if any of the item already exist or not, If it exist then it should not insert that (Possible options need to be in the attribute code should be 'Black, Blue, White, Green, Purple')

### Actual result

1. Magento create duplicate option for the same store even the option already exists. Data in the current option array are ( 'Black, Blue, White, White,  Green, Purple')) check the repetative entry for the option value 'White.'

![image](https://user-images.githubusercontent.com/12829227/42747766-47391122-8910-11e8-90b0-8970e8ce459b.png)

here are the sample code:

For step 1: 

Inject the EaV Attribute Factory in your constructor and perform this action
```
$attributeData = [
                    'attribute_code' => $code,
                    'is_global' => 1,
                    'frontend_label' => $code,
                    'frontend_input' => 'select',
                    'default_value_text' => '',
                    'default_value_yesno' => 0,
                    'default_value_date' => '',
                    'default_value_textarea' => '',
                    'is_unique' => 0,
                    'apply_to' => 0,
                    'is_required' => 0,
                    'is_configurable' => 1,
                    'is_searchable' => 1,
                    'is_comparable' => 1,
                    'is_user_defined' => 1,
                    'is_visible_in_advanced_search' => 1,
                    'is_used_for_price_rules' => 0,
                    'is_wysiwyg_enabled' => 0,
                    'is_html_allowed_on_front' => 1,
                    'is_visible_on_front' => 0,
                    'used_in_product_listing' => 0,
                    'used_for_sort_by' => 1,
                    'is_filterable' => 1,
                    'is_filterable_in_search' => 1,
                    'backend_type' => 'varchar'
                ];
/*** Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory  */ 
                $this->_eavAttributeFactory->create()
                    ->addData($attributeData)
                    ->setEntityTypeId($this->getEntityTypeId('catalog_product'))
                    ->save();
```
Step 2: 
```
        $attribute_code = 'COLOURS';
        $options        =   array('Black', 'Green');

  /*** Magento\Eav\Setup\EavSetup  */
        $this->eavSetupFactory->create()
            ->addAttributeOption(
                [
                    'values' => $options,
                    'attribute_id' => $this->getAttributeIdbyCode($attribute_code) := '4'
                ]
            );
```
Step 3:
```
        $attribute_code = 'COLOURS';
        $options        =   array('Black', 'White');

  /*** Magento\Eav\Setup\EavSetup  */
        $this->eavSetupFactory->create()
            ->addAttributeOption(
                [
                    'values' => $options,
                    'attribute_id' => $this->getAttributeIdbyCode($attribute_code) := '4'
                ]
            );
```

**Comment by magento-engcom-team** [comment]
Hi @zeeshan2523806. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@zeeshan2523806 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by zeeshan2523806** [comment]
its development bug, not sure how i can produce this on the existing setup

**Comment by engcom-backlog-nickolas** [comment]
Hello @zeeshan2523806, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by medmek** [comment]
any workaround for the mean time ?
I see that in contrast with 'value' the 'values' doesn't have an `$intOptionId` by which the replacement could be done easily, so for the mean time I propose to use the sortOrder value.

https://github.com/magento/magento2/blob/24376e9f039d1c4aaa75f3ac9787202a0eaf66b5/app/code/Magento/Eav/Setup/EavSetup.php#L931-L941

line 933  :

    // add option
    $this->setup->getConnection()->delete($optionTable, ['attribute_id =?' =>$option['attribute_id'], 'sort_order =?' => $sortOrder]);
    $data = ['attribute_id' => $option['attribute_id'], 'sort_order' => $sortOrder];
    ...

Haven't had time to test it yet

**Comment by magento-engcom-team** [comment]
Hi @otsmyh. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by magento-engcom-team** [comment]
Hi @zeeshan2523806. Thank you for your report. 
The issue has been fixed in magento/magento2#21424 by @novikor in 2.3-develop branch
Related commit(s):
 - [e79e57be45d798763728e4449443616df07cda5a](https://github.com/magento/magento2/commit/e79e57be45d798763728e4449443616df07cda5a)
 - [a8ec393227a5aeca1ba9aca5c33064cd6e3d9a61](https://github.com/magento/magento2/commit/a8ec393227a5aeca1ba9aca5c33064cd6e3d9a61)

The fix will be available with the upcoming 2.3.4 release.
